datasets.py

```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
+    if not destination_dir.is_dir():
+        assert(destination_dir.parent.is_dir())
+        Path.mkdir(destination_dir)
+
+    all_data = _load_spoken_digit_dataset_tf(percent_load, True, download=download)
+    total_count = len(all_data)
+    print(f'Writing {total_count} examples as numpy arrays')
+    
+    example_id = 0
+    for example_audio, example_label in all_data:
+        destination_path = destination_dir / f'example_{example_id}.npy'
+        audio = example_audio.numpy()
+        label = int(example_label)
+        example_type = np.dtype([('label', np.int8), ('audio', np.float32, (len(audio),))])
+        example_array = np.rec.array((label, audio), dtype=example_type)
+        np.save(destination_path, example_array)
+        example_id += 1
+
+def _get_spoken_digit_feature_channels(feature_name):
+    return 2 if feature_name == 'spectrum_norm' else 1
+
+def _get_spoken_digit_signature_and_padded_shape(config, load_labels):
+    windowed_frame_config = config['dsp']['windowed_frames']
+    frame_size = dsp.get_windowed_frame_size(windowed_frame_config)
+    num_bins = frame_size // 2 + 1
+    
+    stft_features_config = config['dsp']['stft_features']
+    feature_names = stft_features_config['features']
+    
+    signature = {}
+    padded_shapes = {}
+    for feature_name in feature_names:
+        num_channels = _get_spoken_digit_feature_channels(feature_name)
+        signature[feature_name] = tf.TensorSpec(shape=(None, num_bins, num_channels), dtype=tf.float32)
+        padded_shapes[feature_name] = (None, num_bins, num_channels)
+        
+    signature = [signature]
+        
+    if load_labels:
+        signature.append(tf.TensorSpec(shape=(), dtype=tf.int8))
+        padded_shapes = (padded_shapes, ())
+    
+    return tuple(signature), padded_shapes
+
+
+def _extract_spoken_digit_features_npy(example, config, load_labels):
+    audio = example.audio # this should be a numpy array
+    label = example.label if load_labels else None
+    # TODO data augmentation
+
+    windowed_frame_config = config['dsp']['windowed_frames']
+    stft_features_config = config['dsp']['stft_features']
+    
+    windowed_frame_data = dsp.windowed_frame_analysis(audio, windowed_frame_config)
+    features = dsp.stft_analysis(windowed_frame_data['frames'], stft_features_config)
+    
+    # convert numpy arrays to tensors
+    for feature_name in features:
+        num_channels = _get_spoken_digit_feature_channels(feature_name)
+        feature_frames = features[feature_name].view(f'({num_channels},)float')
+        features[feature_name] = tf.convert_to_tensor(feature_frames, dtype=tf.float32)
+    
+    if label is not None:
+        label = tf.convert_to_tensor(label, dtype=tf.int8)
+        return (features, label)
+    
+    return features
+
+
+def _count_spoken_digit_dataset_npy(data_dir):
+    return sum([1 for _ in data_dir.glob('*.npy')])
```

Comment:
`len(data_dir.glob('*.npy')`?

Reply:
I think it would have to be `len(list(data_dir.glob('*.npy')))` unless `len(generator)` is some newer feature I don't know about..

I agree the syntax isn't beautiful but that's why I put it in a function. `len(list)` approach might be slightly less desirable from a memory standpoint in situations where there are many elements, but if you feel strongly about it I can change it, probably doesn't matter in this case